onedal/neighbors/neighbors.py

```diff
@@ -1,578 +1,710 @@
+
+        classes_ = self.classes_
+        _y = self._y
+        if not self.outputs_2d_:
+            _y = self._y.reshape((-1, 1))
+            classes_ = [self.classes_]
+
+        n_queries = _num_samples(X)
+
+        weights = self._get_weights(neigh_dist, self.weights)
+        if weights is None:
+            weights = np.ones_like(neigh_ind)
+
+        all_rows = np.arange(n_queries)
+        probabilities = []
+        for k, classes_k in enumerate(classes_):
+            pred_labels = _y[:, k][neigh_ind]
+            proba_k = np.zeros((n_queries, classes_k.size))
+
+            # a simple ':' index doesn't work right
+            for i, idx in enumerate(pred_labels.T):  # loop is O(n_neighbors)
+                proba_k[all_rows, idx] += weights[:, i]
+
+            # normalize 'votes' into real [0,1] probabilities
+            normalizer = proba_k.sum(axis=1)[:, np.newaxis]
+            normalizer[normalizer == 0.0] = 1.0
+            proba_k /= normalizer
+
+            probabilities.append(proba_k)
+
+        if not self.outputs_2d_:
+            probabilities = probabilities[0]
+
+        return probabilities
+
+    def kneighbors(self, X=None, n_neighbors=None,
+                   return_distance=True, queue=None):
+        return super()._kneighbors(X, n_neighbors, return_distance, queue=queue)
+
+
+class KNeighborsRegressor(NeighborsBase, RegressorMixin):
+    def __init__(self, n_neighbors=5, *,
+                 weights='uniform', algorithm='auto',
+                 p=2, metric='minkowski', metric_params=None, **kwargs):
+        super().__init__(
+            n_neighbors=n_neighbors,
+            algorithm=algorithm,
+            metric=metric, p=p,
+            metric_params=metric_params,
+            **kwargs)
+        self.weights = weights
+
+    def _get_onedal_params(self, data):
+        params = super()._get_onedal_params(data)
+        params['result_option'] = 'responses'
+        return params
+
+    def _get_daal_params(self, data):
+        params = super()._get_daal_params(data)
+        params['resultsToCompute'] = 'computeIndicesOfNeighbors|computeDistances'
+        params['resultsToEvaluate'] = 'none'
+        return params
+
+    def _onedal_fit(self, X, y, queue):
+        gpu_device = queue is not None and queue.sycl_device.is_gpu
+        if self.effective_metric_ == 'euclidean' and not gpu_device:
+            params = self._get_daal_params(X)
+            if self._fit_method == 'brute':
+                train_alg = bf_knn_classification_training
+
+            else:
+                train_alg = kdtree_knn_classification_training
+
+            return train_alg(**params).compute(X, y).model
+
+        policy = _get_policy(queue, X, y)
+        params = self._get_onedal_params(X)
+        train_alg_regr = _backend.neighbors.regression.train
+        train_alg_cls = _backend.neighbors.classification.train
+        train_alg = train_alg_regr if gpu_device else train_alg_cls
```

Comment:
Are we using classification algorithm for CPU device?

Reply:
Search algorithm to be correct. We've used "classification" kernel for computing indices and distances, we have "search" kernel at oneDAL interfaces which are not required to get Y for training stage. I have changed approach since daal4py implementation has made and now we can use "search" explicitly.